mangaki/mangaki/views.py

```diff
@@ -891,17 +893,70 @@ def fix_suggestion(request, suggestion_id):
     return render(request, 'fix/fix_suggestion.html', context)
 
 
+def nsfw_grid(request):
+    #FIXME: remove duplicate suggestions OR group by work ID
+    user = request.user if request.user.is_authenticated else None
+    nsfw_suggestion_list = Suggestion.objects.select_related(
+            'work', 'user', 'work__category').prefetch_related(
+            'work__category', 'evidence_set__user').filter(
+                Q(problem__in=('nsfw', 'n_nsfw')),
+                Q(is_checked=False),
+                ~Q(evidence__user=user) | Q(evidence=None)
+            ).order_by('-work__nb_ratings')
+    count_nsfw_left = nsfw_suggestion_list.count()
+
+    paginator = Paginator(nsfw_suggestion_list, NSFW_GRID_PER_PAGE)
+    page = request.GET.get('page')
+
+    try:
+        suggestions = paginator.page(page)
+    except PageNotAnInteger:
+        suggestions = paginator.page(1)
+    except EmptyPage:
+        suggestions = paginator.page(paginator.num_pages)
+
+    nsfw_states = []
+    supposed_nsfw = []
+    for suggestion in suggestions:
+        supposed_nsfw.append(suggestion.problem == 'nsfw')
+
+        for evidence in suggestion.evidence_set.all():
+            if evidence.user == request.user:
+                agrees_with_problem = evidence.agrees
+                agrees = ((agrees_with_problem and suggestion.problem == 'nsfw')
+                       or (not agrees_with_problem and not suggestion.problem == 'nsfw'))
+                nsfw_states.append(agrees)
+                break
+        else:
+            nsfw_states.append(None)
+
+    suggestions_with_states = list(zip(suggestions, nsfw_states, supposed_nsfw))
+
+    context = {
+        'suggestions_with_states': suggestions_with_states,
+        'suggestions': suggestions,
+        'count_nsfw_left': count_nsfw_left
+    }
+
+    return render(request, 'fix/nsfw_grid.html', context)
+
+
 @login_required
 def update_evidence(request):
-    if request.method != 'POST':
-        redirect('fix-index')
+    if request.method != 'POST' or not request.user.is_authenticated:
+        return redirect('fix-index')
+
+    agrees_values = map(lambda x: x == 'True', request.POST.getlist('agrees'))
```

Comment:
It's a bit of a shame we cannot simplify easily this logic, let's add a comment to explain why we could not use a form or why we didn't use a JSON object or something like this.
That will help the next person which will read this code.